File: api/src/main/kotlin/api/routing/stops/train/cano/ElCanoAuth.kt
package api.routing.stops.train.cano

import api.exceptions.BusTrackerException
import org.apache.commons.codec.digest.MessageDigestAlgorithms
import java.math.BigInteger
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.text.SimpleDateFormat
import java.util.*
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

private const val SLASH = "/"

abstract class ElcanoAuth {
    protected var contentType: String? = null
    protected var elcanoAccessKey: String? = null
    protected var elcanoSecretKey: String? = null
    protected var host: String? = null
    protected var httpMethodName: String? = null
    protected var params: String? = null
    protected var path: String? = null
    protected var payload: String? = null
    protected var requestDate: Date? = null
    private var signedHeaders: String? = null
    protected var xElcanoClient: String? = null
    protected var xElcanoDate: String? = null
    protected var xElcanoDateSimple: String? = null
    protected var xElcanoUserId: String? = null

    fun getHeaders(): Map<String, String?> {
        val headers = TreeMap<String, String?>()
        headers[HEADER_X_ELCANO_HOST] = host
        headers[HEADER_CONTENT_CONTENTTYPE] = contentType
        headers[HEADER_X_ELCANO_CLIENT] = xElcanoClient
        headers[HEADER_X_ELCANO_DATE] = timeStamp
        headers[HEADER_X_ELCANO_USERID] = xElcanoUserId
        headers["Authorization"] = calculateHeaderAuthorization()
        return headers
    }

    private fun calculateHeaderAuthorization(): String {
        val calculateSignature = calculateSignature(prepareStringToSign(prepareCanonicalRequest()))
        return buildAuthorizationString(calculateSignature)
    }

    private fun prepareCanonicalRequest(): String {
        val sb = StringBuilder()
        sb.append(httpMethodName)
        sb.append("\n")
        sb.append(path)
        sb.append("\n")
        sb.append(params)
        sb.append("\n")
        sb.append(HEADER_CONTENT_CONTENTTYPE.lowercase(Locale.getDefault()))
        sb.append(":")
        sb.append(contentType)
        sb.append("\n")
        sb.append(HEADER_X_ELCANO_HOST.lowercase(Locale.getDefault()))
        sb.append(":")
        sb.append(host)
        sb.append("\n")
        sb.append(HEADER_X_ELCANO_CLIENT.lowercase(Locale.getDefault()))
        sb.append(":")
        sb.append(xElcanoClient)
        sb.append("\n")
        sb.append(HEADER_X_ELCANO_DATE.lowercase(Locale.getDefault()))
        sb.append(":")
        sb.append(xElcanoDate)
        sb.append("\n")
        sb.append(HEADER_X_ELCANO_USERID.lowercase(Locale.getDefault()))
        sb.append(":")
        sb.append(xElcanoUserId)
        sb.append("\n")
        sb.append(HEADER_CONTENT_CONTENTTYPE.lowercase(Locale.getDefault()))
        sb.append(";")
        sb.append(HEADER_X_ELCANO_HOST.lowercase(Locale.getDefault()))
        sb.append(";")
        sb.append(HEADER_X_ELCANO_CLIENT.lowercase(Locale.getDefault()))
        sb.append(";")
        sb.append(HEADER_X_ELCANO_DATE.lowercase(Locale.getDefault()))
        sb.append(";")
        sb.append(HEADER_X_ELCANO_USERID.lowercase(Locale.getDefault()))
        sb.append("\n")
        signedHeaders =
            HEADER_CONTENT_CONTENTTYPE.lowercase(Locale.getDefault()) + ";" + HEADER_X_ELCANO_HOST.lowercase(
                Locale.getDefault()
            ) + ";" + HEADER_X_ELCANO_CLIENT.lowercase(Locale.getDefault()) + ";" + HEADER_X_ELCANO_DATE.lowercase(
                Locale.getDefault()
            ) + ";" + HEADER_X_ELCANO_USERID.lowercase(Locale.getDefault())
        val formatPayload = formatPayload(payload)
        payload = formatPayload
        sb.append(toHex(formatPayload))
        return sb.toString()
    }

    private fun prepareStringToSign(str: String): String =
        ("HMAC-SHA256\n$xElcanoDate\n${xElcanoDateSimple}$SLASH${xElcanoClient}$SLASH${xElcanoUserId}$SLASH$ELCANO_REQUEST\n") + toHex(
            str
        )

    private fun calculateSignature(str: String): String = bytesToHex(
        hmacSha256(
            getSignatureKey(
                elcanoSecretKey ?: throw BusTrackerException.InternalServerError("Key is null"),
                xElcanoDateSimple ?: throw BusTrackerException.InternalServerError("Date is null"),
                xElcanoClient ?: throw BusTrackerException.InternalServerError("Client is null")
            ), str
        )
    )

    private fun buildAuthorizationString(str: String): String =
        "${"HMAC-SHA256 Credential=${elcanoAccessKey}$SLASH"}${xElcanoDateSimple}$SLASH${xElcanoClient}$SLASH${xElcanoUserId}$SLASH$ELCANO_REQUEST,SignedHeaders=${signedHeaders},Signature=$str"

    private fun formatPayload(str: String?): String = str?.replace("\r", "")
        ?.replace("\n", "")
        ?.replace(" ", "")
        ?: ""

    private fun toHex(str: String): String {
        val instance = MessageDigest.getInstance(MessageDigestAlgorithms.SHA_256)
        instance.update(str.toByteArray(StandardCharsets.UTF_8))
        return String.format("%064x", *arrayOf<Any>(BigInteger(1, instance.digest())))
    }

    private fun hmacSha256(bArr: ByteArray?, str: String): ByteArray {
        val instance = Mac.getInstance("HmacSHA256")
        instance.init(SecretKeySpec(bArr, "HmacSHA256"))
        return instance.doFinal(str.toByteArray(StandardCharsets.UTF_8))
    }

    private fun getSignatureKey(str: String, str2: String, str3: String): ByteArray {
        return hmacSha256(hmacSha256(hmacSha256(str.toByteArray(StandardCharsets.UTF_8), str2), str3), ELCANO_REQUEST)
    }

    private fun bytesToHex(bArr: ByteArray) = bArr.joinToString("") { "%02x".format(it) }

    private val timeStamp: String get() = getTimeStamp(requestDate)

    companion object {
        protected const val ELCANO_REQUEST: String = "elcano_request"
        const val HEADER_CONTENT_CONTENTTYPE: String = "Content-type"
        const val HEADER_X_ELCANO_CLIENT: String = "X-Elcano-Client"
        const val HEADER_X_ELCANO_DATE: String = "X-Elcano-Date"
        const val HEADER_X_ELCANO_HOST: String = "X-Elcano-Host"
        const val HEADER_X_ELCANO_USERID: String = "X-Elcano-UserId"

        @JvmStatic
        protected fun getTimeStamp(date: Date?): String {
            val simpleDateFormat = SimpleDateFormat("yyyyMMdd'T'HHmmss'Z'")
            simpleDateFormat.timeZone = TimeZone.getTimeZone("UTC")
            return simpleDateFormat.format(date)
        }

        @JvmStatic
        protected fun getDate(date: Date?): String {
            val simpleDateFormat = SimpleDateFormat("yyyyMMdd")
            simpleDateFormat.timeZone = TimeZone.getTimeZone("UTC")
            return simpleDateFormat.format(date)
        }
    }
}

File: MadridTransporte.Api/Clients/Train/ElCanoAuth.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;

namespace MadridTransporte.Api.Clients.Train;

public class ElCanoAuth
{
    private const string ElcanoRequest = "elcano_request";
    private const string HeaderContentType = "Content-type";
    private const string HeaderXElcanoHost = "X-Elcano-Host";
    private const string HeaderXElcanoClient = "X-Elcano-Client";
    private const string HeaderXElcanoDate = "X-Elcano-Date";
    private const string HeaderXElcanoUserId = "X-Elcano-UserId";

    private readonly string _accessKey;
    private readonly string _secretKey;
    private readonly string _contentType;
    private readonly string _host;
    private readonly string _httpMethod;
    private readonly string _params;
    private readonly string _path;
    private readonly string _payload;
    private readonly string _xElcanoClient;
    private readonly string _xElcanoDate;
    private readonly string _xElcanoDateSimple;
    private readonly string _xElcanoUserId;

    public ElCanoAuth(
        string accessKey,
        string secretKey,
        string host,
        string path,
        string httpMethod,
        string contentType,
        string xElcanoClient,
        string xElcanoUserId,
        string payload,
        string queryParams
    )
    {
        _accessKey = accessKey;
        _secretKey = secretKey;
        _host = host;
        _path = path;
        _httpMethod = httpMethod;
        _contentType = contentType;
        _xElcanoClient = xElcanoClient;
        _xElcanoUserId = xElcanoUserId;
        _payload = payload.Replace(" ", "");
        _params = queryParams;

        var now = DateTime.UtcNow;
        _xElcanoDate = now.ToString("yyyyMMdd'T'HHmmss'Z'", CultureInfo.InvariantCulture);
        _xElcanoDateSimple = now.ToString("yyyyMMdd", CultureInfo.InvariantCulture);
    }

    public Dictionary<string, string> GetHeaders()
    {
        var headers = new Dictionary<string, string>
        {
            [HeaderXElcanoHost] = _host,
            [HeaderContentType] = _contentType,
            [HeaderXElcanoClient] = _xElcanoClient,
            [HeaderXElcanoDate] = _xElcanoDate,
            [HeaderXElcanoUserId] = _xElcanoUserId,
            ["Authorization"] = CalculateAuthorization(),
        };
        return headers;
    }

    private string CalculateAuthorization()
    {
        var canonicalRequest = PrepareCanonicalRequest();
        var stringToSign = PrepareStringToSign(canonicalRequest);
        var signature = CalculateSignature(stringToSign);
        return BuildAuthorizationString(signature);
    }

    private string PrepareCanonicalRequest()
    {
        var signedHeaders =
            $"{HeaderContentType.ToLower()};{HeaderXElcanoHost.ToLower()};{HeaderXElcanoClient.ToLower()};{HeaderXElcanoDate.ToLower()};{HeaderXElcanoUserId.ToLower()}";

        var sb = new StringBuilder();
        sb.Append(_httpMethod).Append('\n');
        sb.Append(_path).Append('\n');
        sb.Append(_params).Append('\n');
        sb.Append(HeaderContentType.ToLower()).Append(':').Append(_contentType).Append('\n');
        sb.Append(HeaderXElcanoHost.ToLower()).Append(':').Append(_host).Append('\n');
        sb.Append(HeaderXElcanoClient.ToLower()).Append(':').Append(_xElcanoClient).Append('\n');
        sb.Append(HeaderXElcanoDate.ToLower()).Append(':').Append(_xElcanoDate).Append('\n');
        sb.Append(HeaderXElcanoUserId.ToLower()).Append(':').Append(_xElcanoUserId).Append('\n');
        sb.Append(signedHeaders).Append('\n');

        var formattedPayload = _payload.Replace("\r", "").Replace("\n", "").Replace(" ", "");
        sb.Append(ToHex(formattedPayload));

        return sb.ToString();
    }

    private string PrepareStringToSign(string canonicalRequest) =>
        $"HMAC-SHA256\n{_xElcanoDate}\n{_xElcanoDateSimple}/{_xElcanoClient}/{_xElcanoUserId}/{ElcanoRequest}\n{ToHex(canonicalRequest)}";

    private string CalculateSignature(string stringToSign)
    {
        var signingKey = GetSignatureKey(_secretKey, _xElcanoDateSimple, _xElcanoClient);
        return BytesToHex(HmacSha256(signingKey, stringToSign));
    }

    private string BuildAuthorizationString(string signature)
    {
        var signedHeaders =
            $"{HeaderContentType.ToLower()};{HeaderXElcanoHost.ToLower()};{HeaderXElcanoClient.ToLower()};{HeaderXElcanoDate.ToLower()};{HeaderXElcanoUserId.ToLower()}";
        return $"HMAC-SHA256 Credential={_accessKey}/{_xElcanoDateSimple}/{_xElcanoClient}/{_xElcanoUserId}/{ElcanoRequest},SignedHeaders={signedHeaders},Signature={signature}";
    }

    private static string ToHex(string input)
    {
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(input));
        return Convert.ToHexStringLower(hash);
    }

    private static byte[] HmacSha256(byte[] key, string data)
    {
        using var hmac = new HMACSHA256(key);
        return hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
    }

    private static byte[] GetSignatureKey(string secret, string date, string client)
    {
        var kDate = HmacSha256(Encoding.UTF8.GetBytes(secret), date);
        var kClient = HmacSha256(kDate, client);
        return HmacSha256(kClient, ElcanoRequest);
    }

    private static string BytesToHex(byte[] bytes) => Convert.ToHexStringLower(bytes);
}
